File: src/main/java/cspfj/heuristic/CrossHeuristic.java
package cspfj.heuristic;

import java.lang.reflect.InvocationTargetException;

import cspfj.ParameterManager;
import cspfj.problem.Problem;
import cspfj.problem.Variable;
import cspfj.util.Parameter;

public final class CrossHeuristic implements Heuristic {

	@Parameter("heuristic.variable")
	private static Class<? extends VariableHeuristic> variableHeuristicClass = WDegOnDom.class;

	@Parameter("heuristic.value")
	private static Class<? extends ValueHeuristic> valueHeuristicClass = Lexico.class;

	private final VariableHeuristic variableHeuristic;

	private final ValueHeuristic valueHeuristic;

	static {
		ParameterManager.register(CrossHeuristic.class);
	}

	// private final static Logger logger =
	// Logger.getLogger("cspfj.CrossHeuristic");

	public CrossHeuristic(Problem problem) throws InstantiationException,
			IllegalAccessException, InvocationTargetException,
			NoSuchMethodException, ClassNotFoundException {
		variableHeuristic = variableHeuristicClass
				.getConstructor(Problem.class).newInstance(problem);
		valueHeuristic = valueHeuristicClass.getConstructor().newInstance();
	}

	public Pair selectPair(final Problem problem) {
		final Variable bestVariable = variableHeuristic.selectVariable();
		if (bestVariable == null) {
			return null;
		}
		return new Pair(bestVariable, valueHeuristic.selectIndex(bestVariable));
	}

	public void compute() {
		// logger.fine("Initializing heuristics");
		valueHeuristic.compute();
	}

	public String toString() {
		return "Crossed " + variableHeuristic + ", " + valueHeuristic;
	}
}

File: src/main/scala/cspfj/heuristic/CrossHeuristic.scala
package cspfj.heuristic;

import java.lang.reflect.InvocationTargetException;

import cspfj.Pair;
import cspfj.ParameterManager;
import cspfj.problem.Problem;
import cspfj.problem.Variable;
import cspfj.util.Parameter;

object CrossHeuristic {
  @Parameter("heuristic.variable")
  var variableHeuristicClass: Class[_ <: VariableHeuristic] = classOf[WDegOnDom]

  @Parameter("heuristic.value")
  var valueHeuristicClass: Class[_ <: ValueHeuristic] = classOf[Lexico]

  ParameterManager.register(classOf[CrossHeuristic]);
}

final class CrossHeuristic(problem: Problem) extends Heuristic {

  val variableHeuristic = CrossHeuristic.variableHeuristicClass.getConstructor(classOf[Problem]).newInstance(problem)

  val valueHeuristic = CrossHeuristic.valueHeuristicClass.getConstructor().newInstance()

  def selectPair(problem: Problem) = {
    variableHeuristic.select match {
      case None => None
      case Some(v) => Some(Pair(v, valueHeuristic.selectIndex(v)))
    }
  }

  def compute() {
    // logger.fine("Initializing heuristics");
    valueHeuristic.compute();
  }

  override def toString =
    "Crossed " + variableHeuristic + ", " + valueHeuristic;

}
